Migrate plugin to be dynamic

nazmulidris opened the pull request:
JetBrains has deprecated components to make IDE startup more performant.
The commit migrates the DarkModeSync component into a service and post
startup activity. Here are some links from JetBrains docs about dynamic
plugin migrations.

- https://www.jetbrains.org/intellij/sdk/docs/basics/plugin_structure/dynamic_plugins.html
- https://jetbrains.org/intellij/sdk/docs/basics/plugin_structure/plugin_components.html

nazmulidris committed:
Migrate plugin to be dynamic

JetBrains has deprecated components to make IDE startup more performant.
The commit migrates the DarkModeSync component into a service and post
startup activity. Here are some links from JetBrains docs about dynamic
plugin migrations.

- https://www.jetbrains.org/intellij/sdk/docs/basics/plugin_structure/dynamic_plugins.html
- https://jetbrains.org/intellij/sdk/docs/basics/plugin_structure/plugin_components.html

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeSync.java:
```diff
@@ -26,16 +29,27 @@
+  }
+
+  public static DarkModeSync getInstance(@NotNull Project project) {
+    return project.getService(DarkModeSync.class);
```
This creates a new `DarkModeSync` service for each Project, right? We want one `DarkModeSync` singleton at the application level

nazmulidris replied:
You are right, this will end up creating a service instance for each project. Good catch!

gilday commented on src/main/resources/META-INF/plugin.xml:
```diff
@@ -30,11 +30,10 @@
-    </component>
   </application-components>
   <extensions defaultExtensionNs="com.intellij">
+    <postStartupActivity implementation="com.github.gilday.darkmode.DarkModeSync$MyStartupActivity"/>
```
`postStartupActivity` pertains to the Project lifecycle. Can we use an [AppLifecycleListener](https://upsource.jetbrains.com/idea-ce/file/idea-ce-f0b3942b49880400994f62c064afd4ef909539eb/platform/platform-impl/src/com/intellij/ide/AppLifecycleListener.java) instead? I'm thinking we may need to initialize the singleton in both the `welcomeScreenDisplayed` and `appStarting` callbacks. Thoughts?

nazmulidris replied:
I've migrated `MyStartupActivity` into a `MyLifecycleListener` which implements `AppLifecycleListener`. 

Also `DarkModeSync` no longer implements `Disposable`. Instead it is all handled in the following lifecycle methods.
1. `appStarting()` -> calls `onStartup()`
2. `appWillBeClosed()` -> `calls onClose()`

And there's no longer a need to pass a project to `getInstance()`, which just returns a singleton.

gilday commented on src/main/java/com/github/gilday/darkmode/DarkModeSync.java:
```diff
@@ -26,16 +29,27 @@
 
-  /** @param lafManager IDEA look-and-feel manager for getting and setting the current theme */
-  public DarkModeSync(final LafManager lafManager) {
+  private static final class MyStartupActivity implements StartupActivity.DumbAware {
```
Can we make this listener `public`? I assume `private` works because the platform is using `setAccessible` before using reflection to create a new instance of this listener, but I would rather not rely on that behavior.

nazmulidris replied:
Done!

gilday reviewed (changes requested):
This is awesome! I was unaware of the new APIs like `Service` before reading the links you shared.

One thing I learned is that the `postStartupActivity` listener pertains to the project lifecycle, but Dark Mode Sync operates at the application level. If I understand this correctly, this change creates a new `DarkModeSync` service each time a new project is opened. Instead, we want one `DarkModeSync` service singleton at the Application level.

Before we merge this improvement, let's figure out how to listen for the right [AppLifecycleListener](https://upsource.jetbrains.com/idea-ce/file/idea-ce-f0b3942b49880400994f62c064afd4ef909539eb/platform/platform-impl/src/com/intellij/ide/AppLifecycleListener.java) events to initialize the service as an Application level singleton.

nazmulidris committed:
Migrate plugin to be dynamic

Instead of using StartupActivity, use AppLifecycleListener and make the
DarkModeSync class a singleton.

nazmulidris commented:
Hi Jonathan

I'm glad that you found this PR useful 😄 . JetBrains does make a lot of changes to their platform APIs and it is really difficult to figure out what they are and how that will impact things 😦 .

I've made the changes you requested. I totally missed that a new service instance would be created for every project opened! Great catch 👏 . Please let me know what you think on this commit that should address these changes: https://github.com/nazmulidris/dark-mode-sync-plugin/commit/47323eb5c35c934e702e1e3180e96b9b38377491

Thanks
Nazmul

ChrisCarini commented:
FWIW, I would love to see this PR land! The UX of dynamically loading / unloading plugins is really nice as a user! :)

gilday commented:
Thanks for making those changes! I tested it out tonight and it's not working as expected. I ran `./gradlew runIde` to test manually as I normally do, and the themes didn't update as expected.

I'll dig in and see if I can figure out why
